README.md
# AGNewsNet
Гем для генерации нейронной сети, которая определяет тему поданной на вход новости (Sports, World, Business, Sci/Tec), на датасете AGNews. 
Возможна установка своих параметров обучения - batch_size и count_of_epochs.

# Установка
Установить гем и добавить в GemFile можно следующей командой: 

```python
bundle add AGNewsNet
```

Если bundler не используется для установки зависимостей, гем можно установить следующим образом: 

```python
gem install AGNewsNet
```

# Использование

##### Создание модели
 
Создание модели является обязательным в начале использования гема
При первом создании модели нейросеть будет сохранена в файл agnewsNet.pth

Создание и обучение нейросети с заранее установленными параметрами
(по умолчанию batch_size = 64 и count_of_epochs = 8)

```python
network = AGNews::Net.new
network.createModel
```

Создание и обучение нейросети измененными параметрами 

```python
network = AGNews::Net.new(batch_size = 32, count_of_epochs = 10 )
network.createModel
```

P.S. Часть кода нейросети адаптирована под Ruby с официальной документации PyTorch на Python.
##### Предсказание темы статьи / новости

Если на вход подается файл
```python
network.makePredictionFromFile(FilePath)
```

Если на вход подается строка
```python
network.makePredictionFromString(text)
```
# Тестирование

Для тестирования гема необходимо в терминале выполнить команду
```python
bundle exec rspec
```
 
# Разработка

Запуск тестов осуществляется следующей командой в терминале: 

```python
bundle exec rspec
```

# Лицензия
Гем доступен с открытым исходным кодом в соответствии с условиями [MIT License](https://github.com/Ekaterina0801/AGNewsNet/blob/master/LICENSE )


CHANGELOG.md
# Changelog

## 1.0.0 (28-Oct-22)

* Initial release


spec/agNewsNet_spec.rb
require 'agNewsNet'
require 'rspec'


text1 = "MEMPHIS, Tenn. – Four days ago, Jon Rahm was \t
    enduring the season’s worst weather conditions on Sunday at The \t
    Open on his way to a closing 75 at Royal Portrush, which \t
    considering the wind and the rain was a respectable showing. \t
    Thursday’s first round at the WGC-FedEx St. Jude Invitational \t
    was another story. With temperatures in the mid-80s and hardly any \t
    wind, the Spaniard was 13 strokes better in a flawless round. \t
    Thanks to his best putting performance on the PGA Tour, Rahm \t
    finished with an 8-under 62 for a three-stroke lead, which \t
    was even more impressive considering he’d never played the \t
    front nine at TPC Southwind." #2
text2 = "'Teenage T. rex's monster growth,Tyrannosaurus rex achieved its massive size due to an enormous growth spurt during its adolescent years."#4
text3 = "Fears for T N pension after talks,Unions representing workers at Turner   Newall say they are 'disappointed' after talks with stricken parent firm Federal Mogul." #3
text4 = "West Mulls Boundries for African Fighting (AP),'AP - As the month-end deadline nears for Sudan to disarm the mostly Arab pro-government militias in Darfur, the United Nations and Western powers are in a dilemma over how far to go to stop the killing in an African country."#1
path1 = File.expand_path("../spec/file1.txt", __dir__)
path2 = File.expand_path("../spec/file2.txt", __dir__)
path3 = File.expand_path("../spec/file3.txt", __dir__)
path4 = File.expand_path("../spec/file4.txt", __dir__)

network = AGNews::Net.new
network.createModel

RSpec.describe AGNews do

  it "Test 1" do
    expect(network.makePredictionFromString(text1)).to eq("Sports")
  end

  it "Test 2" do
    expect(network.makePredictionFromString(text2)).to eq("Sci/Tec")
  end

  it "Test 3" do
    expect(network.makePredictionFromString(text3)).to eq("Business")
  end

  it "Test 4" do
    expect(network.makePredictionFromString(text4)).to eq("World")
  end

  it "Test 5" do
    expect(network.makePredictionFromFile(path1)).to eq("Business")
  end

  it "Test 6" do
    expect(network.makePredictionFromFile(path2)).to eq("World")
  end

  it "Test 7" do
    expect(network.makePredictionFromFile(path3)).to eq("Sports")
  end

  it "Test 8" do
    expect(network.makePredictionFromFile(path4)).to eq("Sci/Tec")
  end

end


lib/neuralNetwork.rb
require 'torch-rb'
require 'torchtext'
class Classifier
  attr_writer :batch_size, :count_epochs
  attr_reader :train_dataset, :test_dataset
  attr_writer :model

  def initialize(batch_size=64, count_epochs=5)
    super()
    @batch_size = batch_size
    @count_epochs  = count_epochs
    device = Torch.device(Torch::CUDA.available? ? "cuda" : "cpu")
    @train_dataset,@test_dataset = TorchText::Datasets::AG_NEWS.load(root: ".data", ngrams: 2)
    @model = TextClassifier.new(@train_dataset.vocab.length, 32, 4).to(device)
    @model.load_state_dict(Torch.load("agnewsNet.pth"))
    @model.eval
  end

  class TextClassifier < Torch::NN::Module

    def initialize(vocab_size, embedding_dimension, num_class)
      super()
      lim = 0.5
      @embedding = Torch::NN::EmbeddingBag.new(vocab_size, embedding_dimension, sparse: true)
      @fc = Torch::NN::Linear.new(embedding_dimension, num_class)
      @embedding.weight.data.uniform!(-lim, lim)
      @fc.weight.data.uniform!(-lim, lim)
      @fc.bias.data.zero!
    end

    def forward(text, offsets)
      @fc.call(@embedding.call(text, offsets: offsets))
    end
  end



  def createModel

    #-----------------------Parameters----------------------------
    batch_size = @batch_size # batch size for training
    count_epochs = @count_epochs
    learning_rate = 5
    gamma_for_sheduler = 0.9
    embedding_dimension = 32
    #-------------------------------------------------------------



    #----------------------Preparing dataset-------------------------
    puts "Preparing dataset"
    #train_dataset, test_dataset = TorchText::Datasets::AG_NEWS.load(root: ".data", ngrams: 2)
    train_len = (@train_dataset.length * 0.8).to_i
    split_train, split_test = Torch::Utils::Data.random_split(@train_dataset, [train_len, @train_dataset.length - train_len])
    device = Torch.device(Torch::CUDA.available? ? "cuda" : "cpu")
    #----------------------------------------------------------------

    #-----------------------Definition network-----------------------

    #----------------------------------------------------------------
    vocab_size = @train_dataset.vocab.length
    count_of_classes = @train_dataset.labels.length
    model = TextClassifier.new(vocab_size, embedding_dimension, count_of_classes).to(device)

    criterion = Torch::NN::CrossEntropyLoss.new.to(device)
    optimizer = Torch::Optim::SGD.new(model.parameters, lr: learning_rate)
    scheduler = Torch::Optim::LRScheduler::StepLR.new(optimizer, step_size: 1, gamma: gamma_for_sheduler)

    collate_batch = Proc.new{
      |batch|
      label = Torch.tensor(batch.map { |entry| entry[0] })
      text = batch.map { |entry| entry[1] }
      offsets = [0] + text.map { |entry| entry.size }
      offsets = Torch.tensor(offsets[0..-2]).cumsum(0)
      text = Torch.cat(text)
      [text, offsets, label]
    }

    #----------------------Training network----------------------
    train = Proc.new{
      |split_train|
      data_train = Torch::Utils::Data::DataLoader.new(split_train, batch_size: batch_size, shuffle: true, collate_fn: collate_batch)
      train_loss,train_accuracy = 0,0
      data_train.each_with_index {
        |(text, offsets, label), i|
        optimizer.zero_grad
        text = text.to(device)
        offsets = offsets.to(device)
        label = label.to(device)
        predicted_label = model.call(text, offsets)
        loss = criterion.call(predicted_label, label)
        train_loss += loss.item
        loss.backward
        optimizer.step
        prediction = predicted_label.argmax(1)
        are_eq = prediction.eq(label)
        train_accuracy += are_eq.sum.item
      }
      scheduler.step
      [train_loss / split_train.length, train_accuracy / split_train.length.to_f]
    }
    #--------------------------------------------------------------

    #----------------------Evaluating network----------------------
    evaluate = Proc.new{
      |split_test|
      data_test = Torch::Utils::Data::DataLoader.new(split_test, batch_size: batch_size, collate_fn: collate_batch)
      loss,accuracy = 0,0
      data_test.each do |text, offsets, label|
        text = text.to(device)
        offsets = offsets.to(device)
        label = label.to(device)
        Torch.no_grad do
          predicted_label = model.call(text, offsets)
          loss = criterion.call(predicted_label, label)
          loss += loss.item
          accuracy += predicted_label.argmax(1).eq(label).sum.item
        end
      end
      [loss / split_test.length, accuracy / split_test.length.to_f]
    }
    #--------------------------------------------------------------

    (1..count_epochs).each {
      |epoch|
      start_time = Time.now
      train_loss, train_accuracy = train.call(split_train)
      test_loss, test_accuracy = evaluate.call(split_test)
      final_time = Time.now
      time =  final_time - start_time
      puts "Epoch: %d | time:  %d seconds" % [epoch, time]
      puts "\tTrain Loss: %.4f \t|\tTrain Accuracy: %.1f%% " % [train_loss, train_accuracy * 100]
      puts "\tTest  Loss: %.4f \t|\tTest  Accuracy: %.1f%% " % [test_loss, test_accuracy * 100]
      puts("-------------------------------------------------")
    }

    puts "Checking the results of test dataset"
    test_loss, test_accuracy = evaluate.call(@test_dataset)
    puts "Test accuracy: %.1f%% " % [test_accuracy * 100]
    Torch.save(model.state_dict, "agnewsNet.pth")
    model


  end
end


lib/agNewsNet.rb
require 'torch-rb'
require 'torchtext'
module AGNews
  class Net
    #class << self
    attr_writer :batch_size, :count_epochs
    attr_reader :train_dataset, :test_dataset
    attr_writer :model

    def initialize(batch_size=32, count_epochs=8)
      super()
      @batch_size = batch_size
      @count_epochs  = count_epochs
      device = Torch.device(Torch::CUDA.available? ? "cuda" : "cpu")
      puts "Loading dataset"
      @train_dataset,@test_dataset = TorchText::Datasets::AG_NEWS.load(root: ".data", ngrams: 2)
      #@model = TextClassifier.new(@train_dataset.vocab.length, 32, 4).to(device)
      #@model.load_state_dict(Torch.load("agnewsNet.pth"))
      #@model.eval
    end

    class TextClassifier < Torch::NN::Module

      def initialize(vocab_size, embedding_dimension, num_class)
        super()
        lim = 0.5
        @embedding = Torch::NN::EmbeddingBag.new(vocab_size, embedding_dimension, sparse: true)
        @fc = Torch::NN::Linear.new(embedding_dimension, num_class)
        @embedding.weight.data.uniform!(-lim, lim)
        @fc.weight.data.uniform!(-lim, lim)
        @fc.bias.data.zero!
      end

      def forward(text, offsets)
        @fc.call(@embedding.call(text, offsets: offsets))
      end
    end



    def createModel

      #-----------------------Parameters----------------------------
      batch_size = @batch_size # batch size for training
      count_epochs = @count_epochs
      learning_rate = 5
      gamma_for_sheduler = 0.9
      embedding_dimension = 32
      #-------------------------------------------------------------



      #----------------------Preparing dataset-------------------------
      train_len = (@train_dataset.length * 0.8).to_i
      split_train, split_test = Torch::Utils::Data.random_split(@train_dataset, [train_len, @train_dataset.length - train_len])
      device = Torch.device(Torch::CUDA.available? ? "cuda" : "cpu")
      #----------------------------------------------------------------

      vocab_size = @train_dataset.vocab.length
      count_of_classes = @train_dataset.labels.length
      model = TextClassifier.new(vocab_size, embedding_dimension, count_of_classes).to(device)

      criterion = Torch::NN::CrossEntropyLoss.new.to(device)
      optimizer = Torch::Optim::SGD.new(model.parameters, lr: learning_rate)
      scheduler = Torch::Optim::LRScheduler::StepLR.new(optimizer, step_size: 1, gamma: gamma_for_sheduler)

      collate_batch = Proc.new{
        |batch|
        label = Torch.tensor(batch.map { |entry| entry[0] })
        text = batch.map { |entry| entry[1] }
        offsets = [0] + text.map { |entry| entry.size }
        offsets = Torch.tensor(offsets[0..-2]).cumsum(0)
        text = Torch.cat(text)
        [text, offsets, label]
      }

      #----------------------Training network----------------------
      train = Proc.new{
        |split_train|
        data_train = Torch::Utils::Data::DataLoader.new(split_train, batch_size: batch_size, shuffle: true, collate_fn: collate_batch)
        train_loss,train_accuracy = 0,0
        data_train.each_with_index {
          |(text, offsets, label), i|
          optimizer.zero_grad
          text = text.to(device)
          offsets = offsets.to(device)
          label = label.to(device)
          predicted_label = model.call(text, offsets)
          loss = criterion.call(predicted_label, label)
          train_loss += loss.item
          loss.backward
          optimizer.step
          prediction = predicted_label.argmax(1)
          are_eq = prediction.eq(label)
          train_accuracy += are_eq.sum.item
        }
        scheduler.step
        [train_loss / split_train.length, train_accuracy / split_train.length.to_f]
      }
      #--------------------------------------------------------------

      #----------------------Evaluating network----------------------
      evaluate = Proc.new{
        |split_test|
        data_test = Torch::Utils::Data::DataLoader.new(split_test, batch_size: batch_size, collate_fn: collate_batch)
        loss,accuracy = 0,0
        data_test.each do |text, offsets, label|
          text = text.to(device)
          offsets = offsets.to(device)
          label = label.to(device)
          Torch.no_grad do
            predicted_label = model.call(text, offsets)
            loss = criterion.call(predicted_label, label)
            loss += loss.item
            accuracy += predicted_label.argmax(1).eq(label).sum.item
          end
        end
        [loss / split_test.length, accuracy / split_test.length.to_f]
      }
      #--------------------------------------------------------------
      puts "Training model"
      (1..count_epochs).each {
        |epoch|
        start_time = Time.now
        train_loss, train_accuracy = train.call(split_train)
        test_loss, test_accuracy = evaluate.call(split_test)
        final_time = Time.now
        time =  final_time - start_time
        puts "Epoch: %d | time:  %d seconds" % [epoch, time]
        puts "\tTrain Loss: %.4f \t|\tTrain Accuracy: %.1f%% " % [train_loss, train_accuracy * 100]
        puts "\tTest  Loss: %.4f \t|\tTest  Accuracy: %.1f%% " % [test_loss, test_accuracy * 100]
        puts("-----------------------------------------------------------------")
      }

      puts "Checking the results of test dataset"
      test_loss, test_accuracy = evaluate.call(@test_dataset)
      puts "Test accuracy: %.1f%% " % [test_accuracy * 100]
      Torch.save(model.state_dict, "agnewsNet.pth")
      @model = model
    end
    #end

    def prediction(model, text)
      Torch.no_grad do
        text = Torch.tensor(TorchText::Data::Utils.ngrams_iterator(TorchText::Data::Utils.tokenizer("basic_english").call(text), 2).map { |token| @train_dataset.vocab[token] })
        model.call(text, Torch.tensor([0])).argmax(1).item + 1
      end
    end
    def makePredictionFromString(text)
      labels = {1 => "World",2 => "Sports",3 => "Business",4 => "Sci/Tec"}
      puts "Theme of news: %s" % labels[prediction(@model,text)]
      labels[prediction(@model,text)]
    end

    def makePredictionFromFile(path)
      labels = {1 => "World",2 => "Sports",3 => "Business",4 => "Sci/Tec"}
      f = File.open(path)
      text = f.read
      f.close
      puts "Theme of news: %s" % labels[prediction(@model,text)]
      labels[prediction(@model,text)]
    end
  end
end